port to Qt Creator 6.0

christian-rauch opened the pull request:
~~This is the initial port for [Qt Creator 6 Beta](https://www.qt.io/blog/qt-creator-6-beta-released). I am going to keep this as a draft until it has been officially released. But you can already review and comment now.~~

Port for [Qt Creator 6](https://www.qt.io/blog/qt-creator-6-released).

Notable changes are:
1. switch to Qt6
2. temporary deactivation of the ROS Terminal (qtermwidget) due to missing support for Qt6, I tried to use the Qt5 version but this seem to crash Qt Creator with something related to the Qt5 compatibility layer
3. without qtermwidget, which is not available on Windows, I am also building the Windows plugin in the CI, the plugin itself does not load correctly, presumably because if missing symbols, but having the CI run with MSVC already provides some useful warnings and error messages

christian-rauch committed:
update to Qt Creator 6 with Qt 6.2

christian-rauch committed:
interpret "qt_modules" either as archive name or as package name

christian-rauch committed:
use Utils::FilePath in place of strings

christian-rauch committed:
update QXmlStreamReader API

christian-rauch committed:
update QString API

christian-rauch committed:
update QBoxLayout API

christian-rauch committed:
implement BuildSystem::name()

christian-rauch committed:
update setEnvironment to setEnvironmentChange

christian-rauch committed:
update normalizePathName to normalizedPathName

christian-rauch committed:
port from QRegExp to QRegularExpression

christian-rauch committed:
make qtermwidget5 optional (opt-in)

christian-rauch committed:
concatenate the separate Qt Creator and Qt prefix paths

christian-rauch committed:
correctly set ROSSettings as struct

christian-rauch committed:
build and link static yaml-cpp from source

christian-rauch committed:
remove CORE_EXPORT from RemoveDirectoryDialog

christian-rauch committed:
export the Qt Creator and Qt path separately

christian-rauch committed:
determine the architecture directory of the Qt path automatically

christian-rauch committed:
build qtermwidget on macOS with Qt5 from brew

christian-rauch committed:
enable Windows built with MSVC

christian-rauch committed:
disable ROS Terminal support on CI until qtermwidget supports Qt6

Levi-Armstrong commented on src/project_manager/ros_generic_run_step.cpp:
```diff
@@ -124,6 +125,7 @@ void ROSGenericRunStep::run()
 
   //send roslaunch command
   terminal.sendText(command);
+#endif
```
If this is disabled this none of the run steps will work; is that correct?

christian-rauch replied:
Right, I think that's the implication of disabling the terminal.

Is it possible to use the "Application Output" (the one for I/O for regular executables) instead of the ROS terminal?

Levi-Armstrong replied:
> Right, I think that's the implication of disabling the terminal.

Should we just not build the run steps or does that further complicate things?

christian-rauch replied:
Is everything that can be executed via the ROS plugin (`rosrun`, `roslaunch`, ...) using the `ROSGenericRunStep`? I think in this case it would be better to just use the internal Qt Creator "Application Output" pane (Alt+3). That should be the one that is used for any other executable, e.g. for projects outside of ROS.

Levi-Armstrong replied:
> Is everything that can be executed via the ROS plugin (rosrun, roslaunch, ...) using the ROSGenericRunStep? 

I believe they all do. 

> I think in this case it would be better to just use the internal Qt Creator "Application Output" pane (Alt+3). That should be the one that is used for any other executable, e.g. for projects outside of ROS.

I think this should work and we could provide an option like the debugger to run in the terminal where it launches the system default terminal which should work on both windows, mac, and linux.

christian-rauch replied:
The option to use the system terminal is already in the standard run options. If you check the "Run Settings" for a standard CMake project, it will already provide a check box "Run in terminal" which will use the system terminal. Otherwise, it will use the internal terminal.

Can we inherit from this standard run step, and replace the executable with the `rosrun` and `roslaunch` commands, and just have the nodes and launch files configurable? This would share more code with the Qt Creator plugins, and we could probably remove the "ROS Terminals" entirely.

If independent of ROS, an integrated terminal with bash is desired, there are other terminal plugins for Qt Creator we could refer to, instead of maintaining our own.

Levi-Armstrong replied:
> Can we inherit from this standard run step, and replace the executable with the `rosrun` and `roslaunch` commands, and just have the nodes and launch files configurable? This would share more code with the Qt Creator plugins, and we could probably remove the "ROS Terminals" entirely.

I like this option, just need to make sure it is exported. If not, in the past I have made requests to have them exported without any issues.

christian-rauch replied:
The "Application Output" pane (Alt+3) can actually not be used to provide input via `stdin` to an application, it only provides the output via `stdout`. So I think the only way to have I/O in Qt Creator would be via the external terminal.

However, what is the use case for stdin in `rosrun` and `roslaunch`? Do we really need a proper terminal for running nodes and launch files? If we go with the proposed solution (using the integrated run step), then terminal input would only be supported when a user chooses the external terminal.

At the moment, this PR would effectively inhibit any run steps in `ROSGenericRunStep::run` by just not running them. I currently see the following options:
1. disable all uses of `ROSGenericRunStep` until `qtermwidget` is ported to Qt6
2. manually launch an external terminal in `ROSGenericRunStep`
3. use the standard Qt Creator run steps with the above limitations

Since I am not using `rosrun` or `roslaunch` from Qt Creator, and I expect to switch to ROS2 within a year, I personally would prefer the first solution of just disabling this feature for now. This would also be the one with the least work required.

@Levi-Armstrong What do you suggest how to go on with this?

Levi-Armstrong replied:
I think it may be better to leverage the external terminal until qtermwidget is available. I personally never using Qt Creator for running roslaunch or rosrun, but did use the terminal. I am fine with getting rid of the roslaunch and rosrun because this can be accomplished using the integrated runstep.

christian-rauch replied:
The run steps, that are effectively deactivated by this PR, did not work in earlier versions before. They somehow broke because of API changes when I started porting the plugin to newer versions. So restoring this functionality will take more work than switching from qtermwidget5 to an external terminal. But this is something I would tackle in a dedicated PR.

If you are fine with the other changes, can we go on with merging this?

Levi-Armstrong replied:
If everything can be achieved using native Qt Creator capability the I would push that direction since it reduce the maintenance of this repo. I am fine with removing the roslaunch, rosrun and qtermwidget capability because I feel they most likely were never used.

christian-rauch replied:
I also never used the rosrun etc functionality from the UI. Given that nobody complained that this functionality was missing since I ported the plugin to newer Qt Creator versions, I gather that it actually wasn't used by others either.

If I have the time, I will look into enabling this for ROS2. But the next time the API breaks more than what can be fixed with a search&replace, I am going to remove that functionality entirely from the codebase.

christian-rauch committed:
replace raw pointer by std::shared_ptr for PackageTargetInfo

christian-rauch committed:
check build configuration pointer before access

christian-rauch commented:
@Levi-Armstrong Can you have a look at this PR and suggest a way how we can go forward without the `qtermwidget`?

christian-rauch commented:
@Levi-Armstrong Can you have a look at this? The removal of the `qtermwidget5` effectively disables `ROSGenericRunStep`, but according to https://github.com/ros-industrial/ros_qtc_plugin/issues/459, this is broken anyway for a while now.

I think the "Run Settings" for `rosrun`, `roslaunch`, `ros2 run`, `ros2 launch`, etc. need to be implemented anyway. So we are not losing any features by removing the qtermwidget5.

christian-rauch committed:
remove the runtime dependencies for installation of official plugin archive

The official archive from the CI builds and links 'yaml-cpp' statically. It
is therefore not required anymore at runtime. 'qtermwidget' has been
deactivated for now.
